### PR TITLE
[DSv4] Use BF16 instead of FP32 for indexer score computation

### PR DESCRIPTION
**Pull request by TTThanos**
## Motivation

Switch the sparse MLA indexer's QK score BMM from FP32 to BF16, enabling Tensor Core (WMMA) dispatch instead of FP32 SIMT SGEMM.

    This eliminates:
    - FP32 SIMT SGEMM (cutlass_80_simt_sgemm): 408ms -> 95ms via Tensor Core
    - Associated BF16<->FP32 dtype conversion copies: ~190ms -> 0ms
    
    Benchmarks on 8x RTX PRO 5000 72GB (DeepSeek-V4 Flash, TP8):
    - E2E serving (cc=8, in=4096, out=150): +19% output throughput
    - E2E serving (cc=16): +24% output throughput, -21% TPOT latency
    - Math/reasoning accuracy verified: no degradation

## Accuracy Tests
Use bench_serving to test E2E performance.
image

## Speed Tests and Profiling

Profile clearly display different kernel when changing FP32 to BF16
image

Precision is tested via aime 2026 Math test with fixed seed and temperature.
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #28642682834](https://github.com/sgl-project/sglang/actions/runs/28642682834)
Latest PR Test (Extra): :x: [Run #28642682736](https://github.com/sgl-project/sglang/actions/runs/28642682736)

### CODE REVIEW

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 1]
```diff
@@ -92,14 +81,14 @@ def fp8_paged_mqa_logits_torch(
 
     kv_values_raw = kvcache_gathered[..., :SCALE_OFFSET].contiguous()
     kv_values_fp8 = kv_values_raw.view(dtype=FP8_DTYPE)
-    kv_values = kv_values_fp8.to(torch.float32)
+    kv_values = kv_values_fp8.to(torch.bfloat16)
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/dsv4/indexer.py:84]: Can we add an environ for controlling its datatype.
Also looks like it only takes affect on sm120 hardware?

**Review by Fridge003 — CHANGES_REQUESTED** [review]

**Comment by Fridge003** [comment]
We need to benchmark AIME again. 60% is not expected.

Please benchmark on AIME25 in this way:
When launching models, attach `SGLANG_DEFAULT_THINKING=1 SGLANG_DSV4_THINKING_EFFORT=max`

On client side, use this command:
```
# To install sgl-eval: pip install git+https://github.com/sgl-project/sgl-eval
sgl-eval run aime25 \
  --model deepseek-ai/DeepSeek-V4-Flash --api-key <api-key> \
  --n-repeats 16 --max-tokens 200000 \
  --temperature 1.0 --top-p 1.0 --thinking \
  --out-dir /sgl-workspace/logs \
  --base-url http://localhost:30000/v1
  ```

**Comment by TTThanos** [comment]
> We need to benchmark AIME again. 60% is not expected.
> 
> Please benchmark on AIME25 in this way: When launching models, attach `SGLANG_DEFAULT_THINKING=1 SGLANG_DSV4_THINKING_EFFORT=max`
> 
> On client side, use this command:
> 
> ```
> # To install sgl-eval: pip install git+https://github.com/sgl-project/sgl-eval
> sgl-eval run aime25 \
>   --model deepseek-ai/DeepSeek-V4-Flash --api-key <api-key> \
>   --n-repeats 16 --max-tokens 200000 \
>   --temperature 1.0 --top-p 1.0 --thinking \
>   --out-dir /sgl-workspace/logs \
>   --base-url http://localhost:30000/v1
> ```

No problems, I will make a test again.

**Review by Fridge003 — APPROVED** [review]
